Source/CarthageKit/Project.swift

```diff
@@ -736,20 +739,46 @@ public final class Project {
 
 	/// Attempts to build each Carthage dependency that has been checked out,
 	/// optionally they are limited by the given list of dependency names.
+	/// Cached dependencies whose dependency trees are also cached will not
+	/// be rebuilt unless otherwise specified via build options.
 	///
 	/// Returns a producer-of-producers representing each scheme being built.
 	public func buildCheckedOutDependenciesWithOptions(_ options: BuildOptions, dependenciesToBuild: [String]? = nil, sdkFilter: @escaping SDKFilterCallback = { .success($0.0) }) -> SignalProducer<BuildSchemeProducer, CarthageError> {
 		return loadResolvedCartfile()
 			.flatMap(.merge) { resolvedCartfile in
 				return self.buildOrderForResolvedCartfile(resolvedCartfile, dependenciesToInclude: dependenciesToBuild)
+					.flatMap(.concat) { dependency -> SignalProducer<(Dependency<PinnedVersion>, Set<ProjectIdentifier>), CarthageError> in
+						self.dependencyProjects(for: dependency)
+							.map { projects in (dependency, projects) }
+					}
+					.reduce([]) { (includedDependencies, nextGroup) -> [Dependency<PinnedVersion>] in
+						let (nextDependency, projects) = nextGroup
+						guard options.cacheBuilds else {
+							return includedDependencies + [nextDependency]
+						}
+
+						var dependenciesToBuild = includedDependencies
+						_ = self.shouldSkipBuildForDependency(nextDependency, dependencyProjects: projects, dependenciesToBeBuilt: includedDependencies, platforms: options.platforms)
+							.on(value: { shouldSkip in
+								if !shouldSkip {
+									dependenciesToBuild.append(nextDependency)
+								}
+							})
+							.wait() // We must wait since subsequent calls to reduce rely on the results of the previous project skipping or not
+
+						return dependenciesToBuild
+					}
+					.flatMap(.concat) { dependencies -> SignalProducer<Dependency<PinnedVersion>, CarthageError> in
+						SignalProducer<Dependency<PinnedVersion>, CarthageError>(dependencies)
+					}
```

Comment:
Tried using flatten instead but unfortunately the swift compiler claims there's ambiguity when trying to make the flatten call. And since flatten doesn't take a closure that we can add a type to to nudge the compiler in the right direction, I think we have to stick with the manual flatMap.